server/s3_auth_client.cc

```diff
@@ -675,11 +675,33 @@ bool S3AuthClient::setup_auth_request_body() {
   const char *full_path = request->c_get_full_encoded_path();
   if (full_path != NULL) {
     std::string uri_full_path = full_path;
-    // in encoded uri path space is encoding as '+'
-    // but cannonical request should have '%20' for verification.
-    // decode plus into space, this special handling
-    // is not required for other charcters.
+    // Handle encoding of special characters in uri path space,
+    // if not then cannonical request generation might break with
+    // SignatureDoesNotMatch error.
     S3CommonUtilities::find_and_replaceall(uri_full_path, "+", "%20");
```

Comment:
lets add a function that will replace this in one scan of uri_full_path string, this method will impact performance, as this is called for every request and add to time to first byte.

Reply:
create hashtable  special_char['+'] = "%20", special_char['!'] = "%21" ...
uri_len = uri_full_path.lenght();
string uri_for_canonical;
for(int i=0;i<uri_len;i++){
    if(special_char[uri_full_path[i]]){
        uri_for_canonical += special_char[uri_full_path[i]];
    } else {
       uri_for_canonical += uri_full_path[i];
    }
}

Reply:
@bkirunge7 - here 1 character gets replaced with 3 - so it is string replacement. We are still doing one more optimization. Pls check the newer version once

Reply:
It's one scan, doesnt' depend on replaced char size, we ar appending new chars at end

Reply:
done .. pushed latest changes